CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+    }
+```
+
+## Lowering values
+
+Lowering is very similar to lifting, it's basically just the opposite in various
+directions.
+
+```
+def lower_all(direction, types, values):
+    values = values.iter()
+    types.map(|ty| lower(direction, ty, values))
+
+def lower(direction, type, value):
+  match type
+    f32 => [value as wasm_f32]
+    f64 => [value as wasm_f64]
+    s8 => [value as wasm_i32]
+    u8 => [value as wasm_i32]
+    s16 => [value as wasm_i32]
+    u16 => [value as wasm_i32]
+    s32 => [value as wasm_i32]
+    u32 => [value as wasm_i32]
+    s64 => [value as wasm_i64]
+    u64 => [value as wasm_i64]
+    char => [value as wasm_i32]
+
+    # It's the responsibility of the native language to supply a list where the
+    # in-memory representation of the list matches the canonical ABI.
+    list $ty => {
+      [value.ptr as wasm_i32, value.len as wasm_i32]
+    }
+
+    record $fields => {
+        if is_bitflags($fields):
+          if $fields.len() > 32:
+              [value.bits as wasm_i64]
+          else:
+              [value.bits as wasm_i32]
+        else:
+	  wasm_vals = []
+	  for field in $fields:
+	    wasm_vals.extend(lower(direction, field, value[field.name]))
+          return wasm_vals
+    }
+
+    variant $cases => {
+      match value {
+        the_case(payload)=> {
+          # Recursively lower the optional payload
+          payload_vals = lower(direction, $cases[the_case], payload)
+
+          # Upcast each individual type in the payload to the shared type for
+          # each arm of the variant
+          payload_types = flatten_cases(direction, $cases)
+          for i in 0..len(payload_vals):
+              payload_vals[i] = upcast(payload_vals[i], payload_types[i])
+
+          # If the shared type has more values than this payload generated we
+          # insert zero values to pad
+          for i in len(payload_vals)..len(payload_types)
+              payload_vals.push(zero_value(payload_types[i]))
+
+          [the_case as wasm_i32] + payload_vals
+        }
+      }
+    }
+
+    # It's expected that handles are always represented in a language as a
+    # 32-bit integer, regardless of whether it's defined in the language itself
+    # or a reference to a foreign handle.
+    shared-handle $resource => {
+      [value as wasm_i32]
+    }
+
+    # Buffers being lowered only happens for the parameters of imports, in which
+    # case we're passing a either a pointer/length relative to our own module or
+    a buffer that was passed to us from somewhere else. This is similar to the
+    # `list` case above where the language is responsible for the in-memory
+    # representation of elements in the buffer to match the canonical abi.
+    push-buffer $t | pull-buffer $t => {
+      assert(direction == "import")
+      if value.is_local_buffer:
+        [0 as wasm_i32, value.ptr as wasm_i32, value.len as wasm_32]
+      else:
+        [1 as wasm_i32, value.handle as wasm_i32, 0 as wasm_32]
```

Comment:
How about the idea discussed at the end of the last chat where:
* 0 is guaranteed to be an invalid handle index
* the first `i32` of a local buffer is given `0`
* the first `i32` of a non-local buffer is given the (non-zero) `value.handle`
* non-local buffers also have a `value.view_offset` and `value.view_len` which are normally 0 and the length of the buffer, but can be reduced via some language-specific `slice()`